Update samples and PreMain for Gramine v1.3 compatibility (#346)
edgelesssys/marblerun: docs/docs/building-services/gramine.md

```sh
sudo apt install libsgx-quote-ex-dev
```

## Configuration

### Entrypoint and argv

We provide the `premain-libos` executable with the [MarbleRun Releases](https://github.com/edgelesssys/marblerun/releases). It will contact the Coordinator, set up the environment, and run the actual application.

Set the premain executable as [the entry point](https://gramine.readthedocs.io/en/v1.3/manifest-syntax.html#libos-entrypoint) of the Gramine application and place the actual entry point [in argv0](https://gramine.readthedocs.io/en/v1.3/manifest-syntax.html#command-line-arguments):

```toml
libos.entrypoint = "file:premain-libos"

# argv0 needs to contain the name of your executable
loader.argv = ["hello"]

# add the premain to the list of trusted files
sgx.trusted_files = [
    # ...
    "file:premain-libos"
]
```

After the premain is done running, it will automatically spawn your application.

### Host environment variables

By default, environment variables from the host won't be passed to the application.

> Gramine allows to [pass through select environment variables from the host](https://gramine.readthedocs.io/en/v1.3/manifest-syntax.html#environment-variables).

Reviewer: ```suggestion
Gramine allows to [pass through whitelisted environment variables from the host](https://gramine.readthedocs.io/en/v1.3/manifest-syntax.html#environment-variables).
```

```

### UUID file

The Marble must be able to store its UUID:

```toml
sgx.allowed_files.uuid = "file:uuid"
```

### Remote attestation

The Marble will send an SGX quote to the Coordinator for remote attestation using DCAP attestation:

```toml
sgx.remote_attestation = "dcap"
```

### Enclave size and threads

The premain process is written in Go. The enclave needs to have enough resources for the Go runtime:

```toml
sgx.enclave_size = "1024M"
sgx.thread_num = 16
```

If your application has high memory demands, you may need to increase the size even further.

### Secret files

A Marble's secrets, e.g. a certificate and private key, can be provisioned as files. You can utilize Gramine's in-memory filesystem [`tmpfs`](https://gramine.readthedocs.io/en/latest/manifest-syntax.html#fs-mount-points), so the secrets will never show up on the host's file system :

```toml
fs.mounts = [
    # ...

>     { type = "tmpf", path = "/secrets" },

Reviewer 2: ```suggestion
    { type = "tmpfs", path = "/secrets" },
```
I suppose that's a typo, right?

...
```

> Note that Gramine also allows to store files encrypted on the host's file system. The so encrypted filesystem requires to initialize an encrypted files key by writing it to the virtual `protected_files_key` device.

Reviewer 2: ```suggestion
Note that Gramine also allows to store files encrypted on the host's file system. This is done by providing one or multiple encryption key and defining a mount point as encrypted in combination with the key to be used.
```

Or something simpler, but `protected_files_key` is deprecated and does not need to be mentioned here for the "for debug purpose" case.

Author: Shortened the section by referring to the Gramine docs where possible

Reviewer 2: Generally that's fine, though I would still keep the distinction between the "debug" use and the proper key deployment with MarbleRun. The `insecure` should already give a hint, but since it's important for security better be safe than sorry I guess. 

Also, one might wonder why we even advertise the insecure way. Keeping the distinction between "quick for debug" versus recommendation for production use cases also allows us to show the value MarbleRun provides here (seamless key provisioning).

But that's the last nit I have, don't worry :)

Author: I would prefer to keep our docs on Gramine as minimal as possible, since Gramine's manifest syntax can, and does, change from one release to another.
Users should not need our docs to learn about, or understand Gramine's features.
The fact that encryption keys should be provisioned at runtime by an external application is already mentioned in their docs, and I think they explain it quite well.

Reviewer 2: Oh never mind, I actually didn't fully realize in the diff that you removed the insecure example from the docs.
Sorry, too early in the morning. That's fine then :)

For debugging purposes, you can define the key directly in the Gramine manifest.

```toml

> fs.insecure__keys.[KEY_NAME]] = "[16-byte hex value]"

Reviewer 2: ```suggestion
fs.insecure__keys.[KEY_NAME]] = "[32-character hex value]"
```
I know this existed before, but I would just call it the same way as Gramine does (32 character hex value).

Of course 32 chars hex string == 16-byte hex value but the latter can be slightly ambiguous if you switch between the two docs.

fs.mounts = [
  # ...
  { type = "encrypted", path = "/secrets", uri = "file:/path/to/local/directory", key_name = "[KEY_NAME] },
  # ...
]
```

> MarbleRun allows provisioning the encrypted files key at runtime to your enclave.

Reviewer 2: I would maybe highlight it a bit more that this is actually the recommended way by Gramine and `insecure__keys` is just here for a quick-start and minimal sample.

> For production use cases, MarbleRun can provision your enclave with encrypted file keys at runtime.

Reviewer 2: ```suggestion
For production use cases, MarbleRun can provision your enclave with encrypted file keys at runtime by creating a virtual file under `/dev/attestation/keys/<key_name>` which Graphene can ingest for key provisioning.
```
